### PR TITLE
store: refactor flat_state.rs to reduce amount of annotations

### PR DESCRIPTION
**Pull request by mina86**
Add flat_state::maybe_new function which creates FlatState object when
requested and the feature is enabled.  This removes bunch of
annotations and noise from shard_tries.rs.

Furthermore, split definition of FlateState to one when the feature is
enabled and another when it’s disabled.  This further removes bunch of
annotations this time at definition place.

Furthermore, by making the latter impossible to instantiate (by using
an enum with no variants) this encodes in the type system that the
flat state won’t be created if the feature is not enabled and its
method won’t be called.  This also makes `Option<FlatState>` a ZST.

### CODE REVIEW

File: core/store/src/flat_state.rs [hunk 1]
```diff
@@ -7,58 +7,89 @@
...
+
+    use crate::Store;
+
+    /// Since this has no variants it can never be instantiated.
+    #[derive(Clone)]
+    pub enum FlatState {}
+
+    impl FlatState {
+        pub fn get_ref(&self, _root: &CryptoHash, _key: &[u8]) -> ! {
+            unreachable!()
```
Comment (matklad) [inline core/store/src/flat_state.rs:85]:
```suggestion
            match self {}
```

File: core/store/src/flat_state.rs [hunk 2]
```diff
@@ -7,58 +7,89 @@
...
+    }
 }
+
+#[cfg(not(feature = "protocol_feature_flat_state"))]
+mod imp {
+    use near_primitives::hash::CryptoHash;
+
+    use crate::Store;
+
+    /// Since this has no variants it can never be instantiated.
```
Comment (matklad) [inline core/store/src/flat_state.rs:79]: :heart:

File: core/store/src/flat_state.rs [hunk 3]
```diff
@@ -7,58 +7,89 @@
...
         }
     }
+
+    /// Possibly creates a new [`FlateState`] object backed by given storage.
+    ///
+    /// Always returns `None` if the `protocol_feature_flat_state` Cargo feature is
+    /// not enabled.  Otherwise, returns a new [`FlatState`] object backed by
+    /// specified storage if `use_flat_state` argument is true.
+    pub fn maybe_new(use_flat_state: bool, store: &Store) -> Option<FlatState> {
+        use_flat_state.then(|| FlatState { store: store.clone() })
```
Comment (matklad) [inline core/store/src/flat_state.rs:69]: TBH, not a fan of `bool::then`

**Review by nikurt — APPROVED** [review]

File: core/store/src/flat_state.rs (hunk 3 above)
Comment (mina86) [reply core/store/src/flat_state.rs:69]: I’m not a huge fan either to be honest, but it feels more readable here with one concise line.

File: core/store/src/flat_state.rs (hunk 1 above)
Comment (mina86) [reply core/store/src/flat_state.rs]: You’ve lied to me:

```
error[E0004]: non-exhaustive patterns: type `&FlatState` is non-empty
  --> core/store/src/flat_state.rs:87:19
87 |             match self {}
   |                   ^^^^
   = note: the matched value is of type `&FlatState`
   = note: references are always considered inhabited
```

;)

**Pull request merged**